Bago dti min

arokem opened the pull request:
There are too many commits here, but some of these should go away when you rebase your branch on master. It's the recent merge of the LiFE branch.

The last commit is what I am proposing. How about that? I have set it so that a voxel with 0 signal is considered to be masked (there's something wrong with that one anyway, no?) and a signal with a very low signal has high eigen-values. What do you think? Does this solve your issue? Any other ideas?

arokem committed:
NF: unique rows of an array.

arokem committed:
NF: Added a crutch function. This will need to be removed later on.

And replaced with `move_streamlines`, but I am defering that until I get a
sense of what is happening here.

arokem committed:
NF: Added a first-pass implementation of LiFE.

arokem committed:
RF: Thinking about simplifying this.

Maybe use dipy.sims.voxel directly?

arokem committed:
WIP: Completed the mapping function voxel2fiber, inclduing test.

arokem committed:
Matrix allocation more-or-less working.

arokem committed:
WIP: Still working out how to create this model.

arokem committed:
NF: Added Sparse SGD algorithm for solution of the LiFE equations.

arokem committed:
TST + BF: Make sure that input to np.gradient is an array.

arokem committed:
RF: Move down the fit function.

arokem committed:
RF: docs

arokem committed:
Might not need to isotropic matrix, if we treat that as a mean.

arokem committed:
RF: seems to be running through now. Need more tests.

arokem committed:
TST: Fix typo in test. Reinstate the Fit test, but need something more rigorous

arokem committed:
BF: Need to be a little stricter to be able to round-trip.

arokem committed:
RF: Moving stuff around a bit, and starting to cythonize voxel2fiber

arokem committed:
BF: python 3 compatibility.

arokem committed:
RF: A bit more work on cythonizing this.

arokem committed:
RF: SGD => SPARSE-NNLS

arokem committed:
NF: Class for approximating and caching signals on the sphere.

arokem committed:
BF: Predict with a gtab.

arokem committed:
RF: Be more lenient in testing, and faster in fitting.

arokem committed:
TST: harumph. This needs to work, I think.

arokem committed:
NF: Function to allow multiple constraints (waypoint ROIs) on selection of
streamlines.

arokem committed:
PEP8

arokem committed:
PEP8

arokem committed:
RF: Add numpy 1.8 gradient function in to smoothen the transition to 1.9

http://mail.scipy.org/pipermail/numpy-discussion/2014-October/071334.html

arokem committed:
WIP: example of LiFE based on CSA probabilistic tracking.

arokem committed:
NF: Get T1-weighted data for the Stanford HARDI data-set.

This is resampled and aligned to the DWI data resolution and location.

arokem committed:
RF: This is still not completely thought through (and doesn't work).

Taking it out for now.

arokem committed:
DOC: Show these in the context of the original anatomy.

arokem committed:
DOC: More detail in the reference.

arokem committed:
DOC: First pass of an example of LiFE.

Takes about 5 minutes to run, and more or less fills my laptop's RAM.

arokem committed:
RF: This import is probably not necessary.

Also: label axes.

arokem committed:
BF: Look at the MD5 of the file, not at this function.

arokem committed:
DOC: Some fixes to plotting calls.

arokem committed:
DOC: Add LiFE example to the index.

DOC

arokem committed:
DOC: LiFE example.

MrBago committed:
RF - incremental cleanup of data fetcher

arokem committed:
DOC: space after colons

arokem committed:
Rename the example. Set the response function to a stick.

arokem committed:
RF : sl => streamline

arokem committed:
RF: Several changes to documentation. Also: approx => sphere

arokem committed:
RF: voxel2fiber => voxel2streamline

arokem committed:
PEP8, and some more cleanup.

arokem committed:
RF: Simplified the implementation of `gradient`

arokem committed:
TST: Extend the testing of spdot.

arokem committed:
RF: Moved spdot and sparse_nnls into dipy.core.optimize

arokem committed:
RF: Use dipy.tracking.streamline's transform_streamlines

arokem committed:
TST: Fix the optimization tests (imports, mostly).

arokem committed:
DOC: LiFE example revision.

arokem committed:
TST: Reduce test requirements for spdot (almost equal).

arokem committed:
TST: Reduce test requirements a bit more.

arokem committed:
DOC: Complete this docstring.

arokem committed:
RF: Use xrange when possible/appropriate.

stefanv committed:
Use importlib instead of __import__

arokem committed:
TST: Added end-to-end testing and comparison with the Matlab implementation.

arokem committed:
TST: Move the Matlab test files into `data`.

arokem committed:
RF: Use the sphere method 'find_closest'.

arokem committed:
BF: Add backport of importlib for python 2.6

stefanv committed:
Merge pull request #1 from arokem/stefanv-importlib

ENH: Backport importlib

arokem committed:
DOC: Whitespace.

arokem committed:
DOC: hyphen

MrBago committed:
TST - new DTI test to help develop min_signal handling

arokem committed:
RF: Save some memory here.

arokem committed:
RF: Simplify sparse NNLS fitting a bit more.

Garyfallidis committed:
Merge pull request #419 from arokem/life

LiFE

arokem committed:
BF: Relative import

arokem committed:
BF: Absolute import instead.

arokem committed:
BF: Remove the import that was not in the try clause.

stefanv committed:
Merge pull request #2 from arokem/stefanv-importlib

BF: Relative import

Garyfallidis committed:
Merge pull request #471 from stefanv/importlib

Use importlib instead of __import__

arokem committed:
BF: If all the signal is zero, evals are 0. If all the signal is merely very
low, evals are very high.

MrBago commented:
I don't think this works. You cannot pass a test simply by changing the test :). The whole issue here is that we could, and sometimes do get voxels with zeros.

The case with all zeros should be considered masked, there is nothing else we can really do with that, but I'd prefer not to have to deal with it explicitly inside of the loop, that's a pretty big performance penalty.

The whole point of the other test case was to show the shortcomings of the current code. Yes a voxel with very low signals will have high diffusivity. A dwi signal of 0 should be "very low signal" and give a similar result. By changing the test, you've stopped testing the very case I'm concerned about.

My proposal is to revert the code, and handle this in the TensorModel instead of the fit method. This has the added benefit of reducing code duplication because we do not need handle this repeatedly for each fit_method.

arokem commented:
OK - I am just trying to understand the concern. Will keep hacking on
this...

On Wed, Dec 10, 2014 at 2:10 PM, MrBago notifications@github.com wrote:

> I don't think this works. You cannot pass a test simply by changing the
> test :). The whole issue here is that we could, and sometimes do get voxels
> with zeros.
> 
> The case with all zeros should be considered masked, there is nothing else
> we can really do with that, but I'd prefer not to have to deal with it
> explicitly inside of the loop, that's a pretty big performance penalty.
> 
> The whole point of the other test case was to show the shortcomings of the
> current code. Yes a voxel with very low signals will have high diffusivity.
> A dwi signal of 0 should be "very low signal" and give a similar result. By
> changing the test, you've stopped testing the very case I'm concerned about.
> 
> My proposal is to revert the code, and handle this in the TensorModel
> instead of the fit method. This has the added benefit of reducing code
> duplication because we do not need handle this repeatedly for each
> fit_method.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MrBago/dipy/pull/12#issuecomment-66533645.

arokem commented:
Oh - one more detail: what should the eigenvalue be for those voxels (all
zeros, except the b0 measurements)?

I guess it should be isotropic, so the eigen-vectors is np.eye(3), but what
exactly do we set the eigenvalues to be?

On Wed, Dec 10, 2014 at 2:14 PM, Ariel Rokem arokem@gmail.com wrote:

> OK - I am just trying to understand the concern. Will keep hacking on
> this...
> 
> On Wed, Dec 10, 2014 at 2:10 PM, MrBago notifications@github.com wrote:
> 
> > I don't think this works. You cannot pass a test simply by changing the
> > test :). The whole issue here is that we could, and sometimes do get voxels
> > with zeros.
> > 
> > The case with all zeros should be considered masked, there is nothing
> > else we can really do with that, but I'd prefer not to have to deal with it
> > explicitly inside of the loop, that's a pretty big performance penalty.
> > 
> > The whole point of the other test case was to show the shortcomings of
> > the current code. Yes a voxel with very low signals will have high
> > diffusivity. A dwi signal of 0 should be "very low signal" and give a
> > similar result. By changing the test, you've stopped testing the very case
> > I'm concerned about.
> > 
> > My proposal is to revert the code, and handle this in the TensorModel
> > instead of the fit method. This has the added benefit of reducing code
> > duplication because we do not need handle this repeatedly for each
> > fit_method.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/MrBago/dipy/pull/12#issuecomment-66533645.

MrBago committed:
BF - compute min_signal from whole data set.

arokem committed:
RF: Deal with very small signal, and with all-zero signal.

These two cases should be separated: if s0 is non-zero, and all of the d_sig is
zero, the eigenvalues should be high (3.0 per default, the diffusion of water
in water at 37 deg C). If all of the signals are 0, we set the eigenvalues and
eigenvectors to 0.

arokem commented:
Is this too much?

https://github.com/arokem/dipy/commit/304f9cd0e6f5ce56636d6140abeef731860932e7

It deals with the whole issue up-front, regardless of the fitting
algorithm.

Not all tests are passing, but I think it might be doing the right thing
nevertheless. I think that the test failures are coming from places where
the eigen-values and eigenvectors are set to 0, and previously they might
have been set to nan (?). I have to go now, but I will be back online
tomorrow morning.

On Wed, Dec 10, 2014 at 2:18 PM, Ariel Rokem arokem@gmail.com wrote:

> Oh - one more detail: what should the eigenvalue be for those voxels (all
> zeros, except the b0 measurements)?
> 
> I guess it should be isotropic, so the eigen-vectors is np.eye(3), but
> what exactly do we set the eigenvalues to be?
> 
> On Wed, Dec 10, 2014 at 2:14 PM, Ariel Rokem arokem@gmail.com wrote:
> 
> > OK - I am just trying to understand the concern. Will keep hacking on
> > this...
> > 
> > On Wed, Dec 10, 2014 at 2:10 PM, MrBago notifications@github.com wrote:
> > 
> > > I don't think this works. You cannot pass a test simply by changing the
> > > test :). The whole issue here is that we could, and sometimes do get voxels
> > > with zeros.
> > > 
> > > The case with all zeros should be considered masked, there is nothing
> > > else we can really do with that, but I'd prefer not to have to deal with it
> > > explicitly inside of the loop, that's a pretty big performance penalty.
> > > 
> > > The whole point of the other test case was to show the shortcomings of
> > > the current code. Yes a voxel with very low signals will have high
> > > diffusivity. A dwi signal of 0 should be "very low signal" and give a
> > > similar result. By changing the test, you've stopped testing the very case
> > > I'm concerned about.
> > > 
> > > My proposal is to revert the code, and handle this in the TensorModel
> > > instead of the fit method. This has the added benefit of reducing code
> > > duplication because we do not need handle this repeatedly for each
> > > fit_method.
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/MrBago/dipy/pull/12#issuecomment-66533645.

arokem committed:
BF: Set the zero-d_sig case to have np.eye(3) evecs.

Also, deal with setting of min_signal. This still needs a little bit more
cleanup.

arokem commented:
OK - I believe that the tests should be passing now (they're passing on my laptop). We still need to refactor out all the references to min_signal, outside of the `__init__` and `fit` functions of the TensorModel, because this issue should be dealt with upfront in those places. Thoughts?

MrBago commented:
Not the approach that I would have taken :), but it does solve the problem.
How should we go about merging this?

Bago

arokem committed:
TST: One more test-case; relative signal.

arokem commented:
How about I rebase this stuff on master and make a new PR?

MrBago commented:
Sure, I believe with this approach we should be able to remove all
references to min_singal right?

Bago

On Thu, Dec 11, 2014 at 11:18 AM, Ariel Rokem notifications@github.com
wrote:

> How about I rebase this stuff on master and make a new PR?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MrBago/dipy/pull/12#issuecomment-66672069.

arokem commented:
Yes - we can keep it around at the **init** level, so people can use it if
they want, but it is a bit vestigial.

On Thu, Dec 11, 2014 at 11:21 AM, MrBago notifications@github.com wrote:

> Sure, I believe with this approach we should be able to remove all
> references to min_singal right?
> 
> Bago
> 
> On Thu, Dec 11, 2014 at 11:18 AM, Ariel Rokem notifications@github.com
> wrote:
> 
> > How about I rebase this stuff on master and make a new PR?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/MrBago/dipy/pull/12#issuecomment-66672069.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MrBago/dipy/pull/12#issuecomment-66672384.

arokem committed:
RF: Get rid of the `min_signal` kwarg for DTI.

arokem commented:
I decided to go with the more extreme variant of this. What do you think?

On Thu, Dec 11, 2014 at 11:25 AM, Ariel Rokem arokem@gmail.com wrote:

> Yes - we can keep it around at the **init** level, so people can use it if
> they want, but it is a bit vestigial.
> 
> On Thu, Dec 11, 2014 at 11:21 AM, MrBago notifications@github.com wrote:
> 
> > Sure, I believe with this approach we should be able to remove all
> > references to min_singal right?
> > 
> > Bago
> > 
> > On Thu, Dec 11, 2014 at 11:18 AM, Ariel Rokem notifications@github.com
> > wrote:
> > 
> > > How about I rebase this stuff on master and make a new PR?
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/MrBago/dipy/pull/12#issuecomment-66672069.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/MrBago/dipy/pull/12#issuecomment-66672384.

arokem committed:
RF + TST: Added back min_signal, testing with DTI data (not DSI data!).

DTI requires a b0 measurement. This did not exist in the data that was
previously used for testing. For that reason, the eigenvalues in these tests
were always exceedingly low (set by min_diffusivity, I believe). I believe that
it makes more sense to test the model fitting with DTI data. Single b-value and
including a b0 measurement.

arokem committed:
PEP8

arokem committed:
RF: Move the `min_signal` kwarg to the fit method.

MrBago commented:
@arokem I'd like to talk about this, can you contact me offline?

arokem committed:
RF: simplify this, move setting of `min_signal` to __init__ function.

Almost there - still need to deal with that hard-coded minimum for cases where
all the signal is 0. Need to think about that.

arokem committed:
BF: Pop that sucker. So that that kwarg doesn't get propagated.

And change back the script. Whew!

arokem committed:
Whitespace.

arokem committed:
RF: We don't need numpy's set operations, after all.
